plutus-tx/src/PlutusTx/Data/AssocList.hs

```diff
@@ -0,0 +1,438 @@
+{-# LANGUAGE DerivingStrategies #-}
+{-# LANGUAGE FlexibleContexts   #-}
+{-# LANGUAGE NoImplicitPrelude  #-}
+{-# LANGUAGE ViewPatterns       #-}
+
+module PlutusTx.Data.AssocList (
+  AssocList,
+  lookup,
+  member,
+  insert,
+  delete,
+  singleton,
+  empty,
+  null,
+  toList,
+  toBuiltinList,
+  safeFromList,
+  unsafeFromList,
+  unsafeFromBuiltinList,
+  uncons,
+  unsafeUncons,
+  noDuplicateKeys,
+  all,
+  any,
+  union,
+  unionWith,
+  ) where
+
+import PlutusTx.Builtins qualified as P
+import PlutusTx.Builtins.Internal qualified as BI
+import PlutusTx.IsData qualified as P
+import PlutusTx.Prelude hiding (all, any, null, toList, uncons)
+import PlutusTx.These
+
+import Prelude qualified as Haskell
+
+{- | A list associating keys and values backed by `P.BuiltinData`.
+
+This implementation has the following characteristics:
+
+  * The `P.toBuiltinData` and `P.unsafeFromBuiltinData` operations are no-op.
+  * Other operations are slower than @PlutusTx.AssocMap.Map@, although equality
+    checks on keys can be faster due to `P.equalsData`.
+  * Many operations involve converting the keys and/or values to/from `P.BuiltinData`.
+
+Therefore this implementation is likely a better choice than @PlutusTx.AssocMap.Map@
+if it is part of a data type defined using @asData@, and the key and value types
+have efficient `P.toBuiltinData` and `P.unsafeFromBuiltinData` operations (e.g., they
+are primitive types or types defined using @asData@).
+
+An `AssocList` is considered well-defined if it has no duplicate keys. Most operations
+preserve the definedness of the resulting `AssocList` unless otherwise noted.
+-}
+newtype AssocList k a = AssocList P.BuiltinData
+  deriving stock (Haskell.Eq, Haskell.Show)
+  deriving newtype (Eq)
+
+instance P.ToData (AssocList k a) where
+  {-# INLINEABLE toBuiltinData #-}
+  toBuiltinData (AssocList d) = d
+
+instance P.FromData (AssocList k a) where
+  {-# INLINABLE fromBuiltinData #-}
+  fromBuiltinData = Just . AssocList
+
+instance P.UnsafeFromData (AssocList k a) where
+  {-# INLINABLE unsafeFromBuiltinData #-}
+  unsafeFromBuiltinData = AssocList
+
+{-# INLINEABLE lookup #-}
+-- | Look up the value corresponding to the key.
+-- If the `AssocList` is not well-defined, the result is the value associated with
+-- the left-most occurrence of the key in the list.
+lookup :: forall k a. (P.ToData k, P.UnsafeFromData a) => k -> AssocList k a -> Maybe a
+lookup (P.toBuiltinData -> k) m = case lookup' k (toBuiltinList m) of
```

Comment:
Nit
```suggestion
lookup k m = case lookup' (P.toBuiltinData k) (toBuiltinList m) of
```

Reply:
^ agreed, it's weird that one is a view pattern and the other is a regular function call (not a big deal though).